Switch list implementation to more effective one

ZyX-I opened the pull request:
Pretty sure it will end up being dynamically allocated array, but some things are undecided yet. Have a nice collection of data (in Russian though) for an article regarding list implementations in different languages (VimL, Python, Lua(Jit), Perl, MoarVM (Perl6), Ruby, PHP), most (except for VimL) are using reallocated arrays, but details are very different (especially PHP which shares structures and code with hash table; Lua(Jit) also has different kind of hash/array hybrid).

TODO:

- [x] Replace list implementation with something better.
- [x] Use benchmarks to confirm that new list implementation is actually faster.
- [x] Compare memory footprints on actual configuration with new and old code.
- [ ] Update unit tests.
- [ ] Create new functional and/or unit tests operating on larger lists: specifically, lists which are at least twice and at least four times larger then initial array size.
- [ ] Create new functional and/or unit tests testing GC operating on containers with bigger nesting: some GC errors I did encounter occurred after second reallocation of GC stack vector.

ZyX-I committed:
scripts: Add list log parser

It does not do anything useful with the parsed log yet though.

[ci skip]

ZyX-I committed:
scripts: Dump log as YAML

[ci skip]

ZyX-I committed:
scripts: Actually do analyze statistics

[ci skip]

ZyX-I commented:
Some stats out of @justinmk log:

```YAML
---
allalloc: 8920
allappend: 1262
allendfind: 280
allfind: 1105
allfindcachemiss: 1091
allfreelist: 8907
allknown_alloc_int: 0
allmov: 0
allremov: 0
allrevers: 59
alls10init: 0
allsinit: 9904
allsort: 60
allstartfind: 317
allstatic_alloc: 9904
allunknown_alloc: 6
append: {avg: 0.0668503019387647, len: 18878, max: 41, med: 0.0, min: 0}
drop: {avg: 0.07193558639686408, len: 18878, max: 59, med: 0.0, min: 0}
dropsiz: {avg: 1.0, len: 1358, max: 1, med: 1.0, min: 1}
find: {avg: 0.05853374298124801, len: 18878, max: 13, med: 0.0, min: 0}
findcachediff: {avg: 0.8816568047337278, len: 169, max: 1, med: 1, min: -1}
findcachemiss: {avg: 0.05779213899777519, len: 18878, max: 6, med: 0.0, min: 0}
findenddiff: {avg: 4.664253393665159, len: 1105, max: 8, med: 8, min: 0}
findstartdiff: {avg: 0.7809954751131222, len: 1105, max: 4, med: 1, min: 0}
first: {avg: 1.0343256700921708, len: 18878, max: 91, med: 0.0, min: 0}
freecont: {avg: 8907.0, len: 1, max: 8907, med: 8907, min: 8907}
freecontsiz: {avg: 2.484562703491636, len: 8907, max: 53, med: 2, min: 0}
freelistsiz: {avg: 0.0, len: 8907, max: 0, med: 0, min: 0}
insert: {avg: 0.0, len: 18878, max: 0, med: 0.0, min: 0}
iter: {avg: 0.24811950418476533, len: 18878, max: 66, med: 0.0, min: 0}
iterconst: {avg: 0.00031783027863121094, len: 18878, max: 1, med: 0.0, min: 0}
known_alloc_len: {avg: 2.1035372144436257, len: 2714, max: 100, med: 2.0, min: 0}
last: {avg: 0.0, len: 18878, max: 0, med: 0.0, min: 0}
len: {avg: 1.3253522618921496, len: 18878, max: 92, med: 0.0, min: 0}
mov: {avg: 0.0, len: 18878, max: 0, med: 0.0, min: 0}
remov: {avg: 0.0, len: 18878, max: 0, med: 0.0, min: 0}
revers: {avg: 0.0031253310732069074, len: 18878, max: 1, med: 0.0, min: 0}
sort: {avg: 0.003178302786312109, len: 18878, max: 1, med: 0.0, min: 0}
startinsert: {avg: 0.0, len: 18878, max: 0, med: 0.0, min: 0}
unknown_alloc_len: {avg: 2.856670341786108, len: 5442, max: 53, med: 1.0, min: 0}
unsure_alloc_len: {avg: 2.503957783641161, len: 758, max: 5, med: 3.0, min: 0}
```

Full output: https://gist.github.com/e2745696b5c5c60e4d694fbb51c9659e.

ZyX-I commented:
Updated after latest commit:

```YAML
---
allalloc: 8920
allappend: 1262
allendfind: 280
allfind: 1105
allfindcachemiss: 1091
allfreelist: 8907
allknown_alloc_int: 0
allmov: 0
allremov: 0
allrevers: 59
alls10init: 0
allsinit: 9904
allsort: 60
allstartfind: 317
allstatic_alloc: 9904
allunknown_alloc: 6
append: {avg: 0.0668503019387647, len: 18878, max: 41, med: 0.0, min: 0}
drop: {avg: 0.07193558639686408, len: 18878, max: 59, med: 0.0, min: 0}
dropsiz: {avg: 1.0, len: 1358, max: 1, med: 1.0, min: 1}
find: {avg: 0.05853374298124801, len: 18878, max: 13, med: 0.0, min: 0}
findcachediff: {avg: 0.8816568047337278, len: 169, max: 1, med: 1, min: -1}
findcachemiss: {avg: 0.05779213899777519, len: 18878, max: 6, med: 0.0, min: 0}
findenddiff: {avg: 4.664253393665159, len: 1105, max: 8, med: 8, min: 0}
findstartdiff: {avg: 0.7809954751131222, len: 1105, max: 4, med: 1, min: 0}
first: {avg: 1.0343256700921708, len: 18878, max: 91, med: 0.0, min: 0}
freecont: {avg: 8907.0, len: 1, max: 8907, med: 8907, min: 8907}
freecontsiz: {avg: 2.484562703491636, len: 8907, max: 53, med: 2, min: 0}
freelistsiz: {avg: 0.0, len: 8907, max: 0, med: 0, min: 0}
insert: {avg: 0.0, len: 18878, max: 0, med: 0.0, min: 0}
iter: {avg: 0.24811950418476533, len: 18878, max: 66, med: 0.0, min: 0}
iterconst: {avg: 0.00031783027863121094, len: 18878, max: 1, med: 0.0, min: 0}
known_alloc_len: {avg: 2.1035372144436257, len: 2714, max: 100, med: 2.0, min: 0}
last: {avg: 0.0, len: 18878, max: 0, med: 0.0, min: 0}
len: {avg: 1.3253522618921496, len: 18878, max: 92, med: 0.0, min: 0}
lengthdiff: {avg: 2.6475207538703165, len: 8914, max: 100, med: 2.0, min: 0}
maxlength: {avg: 1.2853883701188455, len: 18848, max: 100, med: 0.0, min: 0}
minlength: {avg: 0.003477675566524568, len: 8914, max: 5, med: 0.0, min: 0}
mov: {avg: 0.0, len: 18878, max: 0, med: 0.0, min: 0}
remov: {avg: 0.0, len: 18878, max: 0, med: 0.0, min: 0}
revers: {avg: 0.0031253310732069074, len: 18878, max: 1, med: 0.0, min: 0}
sort: {avg: 0.003178302786312109, len: 18878, max: 1, med: 0.0, min: 0}
startinsert: {avg: 0.0, len: 18878, max: 0, med: 0.0, min: 0}
unknown_alloc_len: {avg: 2.856670341786108, len: 5442, max: 53, med: 1.0, min: 0}
unsure_alloc_len: {avg: 2.503957783641161, len: 758, max: 5, med: 3.0, min: 0}
```

My “typical editing session” (though it yields 69 MiB log after something like five minutes) stats:

```YAML
---
allalloc: 31952
allappend: 31387
allendfind: 10840
allfind: 22234
allfindcachemiss: 16109
allfreelist: 25144
allknown_alloc_int: 361
allmov: 722
allremov: 0
allrevers: 0
alls10init: 0
allsinit: 28976
allsort: 1
allstartfind: 14108
allstatic_alloc: 28976
allunknown_alloc: 13
append: {avg: 0.5151405734543485, len: 60929, max: 870, med: 0, min: 0}
drop: {avg: 0.6805133844310591, len: 60929, max: 869, med: 0, min: 0}
dropsiz: {avg: 1.0278320430263126, len: 41463, max: 98, med: 1, min: 1}
find: {avg: 0.3649165422048614, len: 60929, max: 1739, med: 0, min: 0}
findcachediff: {avg: 0.484904738641915, len: 10235, max: 97, med: 0, min: -29}
findcachemiss: {avg: 0.2643896994862873, len: 60929, max: 1337, med: 0, min: 0}
findenddiff: {avg: 4.397994063146532, len: 22234, max: 117, med: 1.0, min: 0}
findstartdiff: {avg: 1.8097958082216425, len: 22234, max: 97, med: 0.0, min: 0}
first: {avg: 4.671896797912324, len: 60929, max: 3702, med: 0, min: 0}
freecont: {avg: 25144.0, len: 1, max: 25144, med: 25144, min: 25144}
freecontsiz: {avg: 5.380806554247534, len: 25144, max: 1542, med: 1.0, min: 0}
freelistsiz: {avg: 0.0, len: 25144, max: 0, med: 0.0, min: 0}
insert: {avg: 0.032644553496692874, len: 60929, max: 795, med: 0, min: 0}
iter: {avg: 0.15700241264422524, len: 60929, max: 80, med: 0, min: 0}
iterconst: {avg: 0.0033317467872441696, len: 60929, max: 1, med: 0, min: 0}
known_alloc_len: {avg: 15.274523007856342, len: 8910, max: 1542, med: 2.0, min: 0}
last: {avg: 0.0, len: 60929, max: 0, med: 0, min: 0}
len: {avg: 3.2614846788885425, len: 60929, max: 7400, med: 1, min: 0}
lengthdiff: {avg: 5.580937401698647, len: 31790, max: 1542, med: 1.0, min: 0}
maxlength: {avg: 3.271500787815126, len: 60928, max: 1542, med: 0.0, min: 0}
minlength: {avg: 0.35460836741113555, len: 31790, max: 49, med: 0.0, min: 0}
mov: {avg: 0.011849858031479263, len: 60929, max: 160, med: 0, min: 0}
remov: {avg: 0.0, len: 60929, max: 0, med: 0, min: 0}
revers: {avg: 0.0, len: 60929, max: 0, med: 0, min: 0}
sort: {avg: 1.6412545749971278e-05, len: 60929, max: 1, med: 0, min: 0}
startinsert: {avg: 0.017413711040719527, len: 60929, max: 635, med: 0, min: 0}
unknown_alloc_len: {avg: 0.9374942209893666, len: 10815, max: 58, med: 0, min: 0}
unsure_alloc_len: {avg: 1.9331852358168147, len: 11704, max: 93, med: 2.0, min: 0}
```

ZyX-I commented:
It appears that typical list is small and immutable - hardly a surprise. What is a surprise is that at least half of the lists are empty ones. As for operations: `:unlet` is used literally never for deleting list elements, `[]` is hardly ever used for long lists and in any case normally used to get first or last list elements, `remove()` is almost never used with range, @justinmk has something which uses reverse() and sort() relatively often (59 and 60 hits) while I have no plugins calling reverse() at all normally and only one sort() call.

ZyX-I commented:
In case somebody wants to play with my log:


[Original log, comprossed: zyx-list.log.xz.not.gz](https://github.com/neovim/neovim/files/1669685/zyx-list.log.xz.not.gz)

[Resulting YAML file (first section tends to quickly go out of date now though): zyx-list.yml.xz.not.gz](https://github.com/neovim/neovim/files/1669686/zyx-list.yml.xz.not.gz)

ZyX-I commented:
I am thinking of implementing list on top of `kvec_withinit_t(typval_T, 4)`. Should increase memory usage then, but 90% of lists will need exactly one allocation.

justinmk commented:
What are the benefits?

ZyX-I commented:
Of what exactly? Using vector over linked list should be faster (well, except for empty lists), and avoiding second allocation is going to speed up things a bit further for most lists. The vector allocation pattern is though merely “double vector size always”, yet a number of languages (like lua or perl, though first is special case (has *two* vectors to grow separately with some logic behind whether item with numeric key is going to stored in hash or actual array) and second is a bit less then doubles: actually, old_len + min_len - 2; and apparently whoever wrote kvec.h was OK with that) manage to work fine with that.

ZyX-I commented:
By the way, it appears that when using vectors behaviour of `:for` is going to change: previously `sort()` and `reverse()` on the list iterated over caused to continue with the same item, no matter where it appeared after the operation, now it is far easier to code if it continued with the same *index*.

justinmk commented:
> Should increase memory usage then, but 90% of lists will need exactly one allocation.

Now I see, here you were talking about the "double vector size always" pattern.

ZyX-I commented:
@justinmk I was talking about allocating a small array as a part of the list structure always, specifically the second argument to `kvec_withinit_t` (4). Without that memory usage will be *reduced*: vector has size, capacity and a pointer, current list implementation has pointers to first, last and cached elements and also size and cached element index, not to mention two pointers per list element and possible losses due to memory fragmentation and/or alignment issues as lists forced many small allocations, one per each element. With this array it will allocate those four elements no matter what, and they are not going to be used for large lists anyway.

ZyX-I commented:
Of what I checked so far only Ruby somewhat pays this price for its arrays, but it pays with three elements per array (why not four?), and they are of pointer integer type and not `typval_T` like we have.

ZyX-I committed:
eval/typval: Switch list implementation to kvec_withinit_t(typval_T, 4)

ZyX-I committed:
hashtab: Allow using HASHTAB_ITER for cycles which are being modified

Iterating with todo may easily produce crash if items are removed from
dictionary during iteration.

ZyX-I committed:
eval/typval: Refactor tv_clear() so that it does not use typval_encode

ZyX-I committed:
eval: When clearing vimvars do clear dictionaries as well

This caused memory leak when clearing v:msgpack_types which was gone unnoticed 
until now for some weird reason.

ZyX-I committed:
eval/typval_encode: Use kv_drop in place of (void)kv_pop

ZyX-I commented:
Current status: functional tests `eval/*` are working, `legacy/055*` is crashing (specifically, on “list and dictionary types a:000 function argument”), AFAIR all functional tests up to `legacy/055*` were working as well. Benchmarking was not done (deferred until zero crashes on functional tests), unit tests were not refactored. If I am not mistaking, new tv_clear implementation is rather stable now (and it is far easier to debug in any case).

ZyX-I committed:
eval: Do not access list item after removing it

ZyX-I commented:
All functional tests pass locally now; and nvim works normally with my actual config only crashing at exit with `-DEXITFREE` (nothing new, was doing that before the PR, but I did not debug the issue yet). Time for some benchmarks now (not today though).

ZyX-I commented:
Looks like preallocating four elements per list was a very bad idea:

before (`pps nvim -o vsize,size,rss,size,sz`):

```
                                     VSZ  SIZE   RSS  SIZE    SZ
 2678 bin/nvim                     82212 33364 17888 33364 20553
```

after:

```
pps nvim -o vsize,size,rss,size,sz
                                     VSZ  SIZE   RSS  SIZE    SZ
 6558 bin/nvim                    4259928 4211272 1065400 4211272 1064982
```

(these are memory footprints of nvim running with my config on this branch and on master where by “master” I mean the commit I branched of).

ZyX-I committed:
eval/typval: Do not preallocate four list elements

ZyX-I commented:
This is strange, removing four preallocated elements actually doubled vsize in place of reducing it:

```
                                     VSZ  SIZE   RSS  SIZE    SZ
22016 bin/nvim                    8454228 8405572 2113344 8405572 2113557
```

ZyX-I committed:
eval/typval: Also save capacity to log

ZyX-I committed:
kvec: Do not double vec size when it is needed to insert one element

ZyX-I commented:
Appears to be a bug in kv_insert. New footprint:

```
                                     VSZ  SIZE   RSS  SIZE    SZ
 1939 bin/nvim                     82008 33352 16888 33352 20502
```

: this one is completely fine.

ZyX-I committed:
Revert "eval/typval: Do not preallocate four list elements"

This reverts commit fb7057bfdf9ff413f4a5133a38de61e24d83810f.

ZyX-I commented:
And nothing really changes when switching used vector back to the one with four preallocated elements:

```
                                     VSZ  SIZE   RSS  SIZE    SZ
 7302 bin/nvim                     82004 33348 17104 33348 20501
```

ZyX-I committed:
benchmarks: Add some list benchmarks

ZyX-I commented:
Benchmark suite runs two times slower with old list implementation (due to allocations obviously). Results of running actual benchmarks and not the suite:

- Allocation is a bit faster on smaller lists, starting from 7 elements new is always faster.
- Pushing to end is predictably faster with old variant, but not by a significant amount until allocation happens: O(1) amortized in action.
- Inserting to start is predictably faster with old variant, mostly significantly, especially when there is reallocation.
- Inserting into the middle on new implementation happens to beat old one in many cases. Looks like an effect of O(1) indexing with better cache locality.
- Popping element from the end is normally faster with new implementation, unless that triggers a reallocation. Again O(1) amortized.
- Removing element from the start for some reason is faster with new implementation and small arrays. Most likely because of better cache locality and the fact that my code does not bother to downsize vector unless that frees at least 16 elements.
- Removing many (17 in this case) elements at once is always significantly faster with new implementation unless that triggers a reallocation.
- Extending list with itself is again faster with new implementation unless that triggers two reallocations now.

Actual benchmark results:

```
Test    Len    New        Old        Old/New
=============================================
alloc     0:   0.081634   0.069535   0.851790
alloc     1:   0.080465   0.072529   0.901373
alloc     2:   0.085668   0.080209   0.936277
alloc     3:   0.086522   0.083916   0.969880
alloc     4:   0.083089   0.087910   1.058022
alloc     5:   0.103029   0.097396   0.945326
alloc     7:   0.109279   0.111003   1.015776
alloc     8:   0.104954   0.116851   1.113354
alloc     9:   0.107419   0.121110   1.127454
alloc    15:   0.114086   0.160800   1.409463
alloc    16:   0.118409   0.157043   1.326276
alloc    17:   0.127829   0.159443   1.247315
alloc    18:   0.129016   0.167033   1.294669
alloc   127:   0.289720   0.746522   2.576702
alloc   128:   0.297755   0.716826   2.407436
alloc   129:   0.397070   0.724452   1.824494
alloc   255:   0.495239   1.335472   2.696621
alloc   256:   0.492916   1.351563   2.741974
alloc   257:   1.037850   1.357370   1.307867
1->$      0:   0.084100   0.092128   1.095458
1->$      1:   0.085522   0.081237   0.949896
1->$      2:   0.083017   0.085475   1.029608
1->$      3:   0.087682   0.083265   0.949625
1->$      4:   0.100901   0.084302   0.835492
1->$      5:   0.087115   0.082050   0.941858
1->$      7:   0.085293   0.084534   0.991101
1->$      8:   0.115907   0.089820   0.774932
1->$      9:   0.086271   0.080445   0.932469
1->$     15:   0.085262   0.080751   0.947092
1->$     16:   0.131703   0.082737   0.628209
1->$     17:   0.085359   0.087956   1.030424
1->$     18:   0.088464   0.084997   0.960809
1->$    127:   0.087548   0.093593   1.069048
1->$    128:   0.277443   0.090451   0.326017
1->$    129:   0.093736   0.089424   0.953998
1->$    255:   0.092373   0.093989   1.017494
1->$    256:   0.516994   0.089202   0.172540
1->$    257:   0.092933   0.091015   0.979361
1->^      0:   0.089013   0.086721   0.974251
1->^      1:   0.094533   0.106327   1.124761
1->^      2:   0.093021   0.098660   1.060621
1->^      3:   0.093392   0.090215   0.965982
1->^      4:   0.112701   0.089086   0.790463
1->^      5:   0.095664   0.093570   0.978111
1->^      7:   0.093191   0.094701   1.016203
1->^      8:   0.124887   0.089392   0.715783
1->^      9:   0.097626   0.097264   0.996292
1->^     15:   0.096753   0.094702   0.978802
1->^     16:   0.147608   0.094362   0.639274
1->^     17:   0.104164   0.097430   0.935352
1->^     18:   0.106348   0.092594   0.870670
1->^    127:   0.135555   0.100846   0.743949
1->^    128:   0.297815   0.099471   0.334003
1->^    129:   0.134291   0.102929   0.766462
1->^    255:   0.182377   0.103619   0.568158
1->^    256:   0.539013   0.099830   0.185209
1->^    257:   0.171107   0.099292   0.580292
1->/2     2:   0.101425   0.097330   0.959625
1->/2     3:   0.100624   0.107525   1.068582
1->/2     4:   0.115910   0.099099   0.854965
1->/2     5:   0.102822   0.101363   0.985810
1->/2     7:   0.103044   0.107291   1.041215
1->/2     8:   0.135642   0.103622   0.763937
1->/2     9:   0.106591   0.112058   1.051290
1->/2    15:   0.105887   0.117631   1.110911
1->/2    16:   0.149341   0.116983   0.783328
1->/2    17:   0.106932   0.116236   1.087009
1->/2    18:   0.110437   0.117470   1.063683
1->/2   127:   0.131598   0.159312   1.210596
1->/2   128:   0.302475   0.161479   0.533859
1->/2   129:   0.128260   0.162679   1.268353
1->/2   255:   0.151183   0.203322   1.344873
1->/2   256:   0.549391   0.203464   0.370345
1->/2   257:   0.148198   0.206987   1.396692
1<-$      1:   0.158996   0.155913   0.980610
1<-$      2:   0.158457   0.167558   1.057435
1<-$      3:   0.158493   0.162502   1.025294
1<-$      4:   0.166323   0.161417   0.970503
1<-$      5:   0.160397   0.169645   1.057657
1<-$      7:   0.158625   0.168083   1.059625
1<-$      8:   0.159461   0.174093   1.091759
1<-$      9:   0.164687   0.178352   1.082976
1<-$     15:   0.167743   0.166455   0.992322
1<-$     16:   0.158647   0.167753   1.057398
1<-$     17:   0.160052   0.181529   1.134188
1<-$     18:   0.157695   0.173732   1.101696
1<-$    127:   0.163786   0.188972   1.153774
1<-$    128:   0.163263   0.189735   1.162143
1<-$    129:   0.327609   0.192102   0.586376
1<-$    255:   0.168089   0.190178   1.131413
1<-$    256:   0.164597   0.195810   1.189633
1<-$    257:   0.349162   0.192061   0.550063
1<-^      1:   0.154249   0.158860   1.029893
1<-^      2:   0.154465   0.160282   1.037659
1<-^      3:   0.159162   0.168571   1.059116
1<-^      4:   0.155482   0.169259   1.088608
1<-^      5:   0.155916   0.160965   1.032383
1<-^      7:   0.158645   0.163036   1.027678
1<-^      8:   0.159392   0.166298   1.043327
1<-^      9:   0.170784   0.163218   0.955698
1<-^     15:   0.161006   0.169182   1.050781
1<-^     16:   0.161752   0.166066   1.026670
1<-^     17:   0.166481   0.167234   1.004523
1<-^     18:   0.163195   0.161174   0.987616
1<-^    127:   0.191378   0.186840   0.976288
1<-^    128:   0.193322   0.196235   1.015068
1<-^    129:   0.346501   0.189340   0.546434
1<-^    255:   0.231959   0.197649   0.852086
1<-^    256:   0.245862   0.196433   0.798956
1<-^    257:   0.417629   0.195900   0.469077
17<-$    17:   0.230481   0.307486   1.334106
17<-$    18:   0.237909   0.347083   1.458890
17<-$   127:   0.223687   0.342248   1.530031
17<-$   128:   0.220469   0.366888   1.664125
17<-$   129:   0.398192   0.352387   0.884968
17<-$   255:   0.224110   0.359098   1.602329
17<-$   256:   0.223430   0.337153   1.508987
17<-$   257:   0.411851   0.348068   0.845131
17<-^    17:   0.219178   0.295807   1.349620
17<-^    18:   0.218156   0.293497   1.345354
17<-^   127:   0.243635   0.324613   1.332374
17<-^   128:   0.242134   0.324176   1.338829
17<-^   129:   0.378098   0.334093   0.883615
17<-^   255:   0.277846   0.334363   1.203411
17<-^   256:   0.288206   0.328798   1.140844
17<-^   257:   0.482049   0.333770   0.692398
l->l      0:   0.099660   0.091930   0.922436
l->l      1:   0.102327   0.106338   1.039198
l->l      2:   0.100023   0.125413   1.253842
l->l      3:   0.121464   0.118690   0.977162
l->l      4:   0.116236   0.126008   1.084070
l->l      5:   0.142339   0.134431   0.944442
l->l      7:   0.148459   0.142540   0.960130
l->l      8:   0.143609   0.150786   1.049976
l->l      9:   0.159745   0.152934   0.957363
l->l     15:   0.167233   0.189268   1.131762
l->l     16:   0.166215   0.208767   1.256006
l->l     17:   0.203642   0.208727   1.024970
l->l     18:   0.201232   0.211478   1.050916
l->l    127:   0.401648   0.868971   2.163514
l->l    128:   0.405602   0.812365   2.002862
l->l    129:   0.604916   0.834751   1.379945
l->l    255:   0.792159   1.527064   1.927724
l->l    256:   0.775091   1.505702   1.942613
l->l    257:   2.710158   1.517055   0.559766
m->l      0:   0.210829   0.224572   1.065186
m->l      1:   0.219544   0.229523   1.045453
m->l      2:   0.212841   0.244532   1.148895
m->l      3:   0.225780   0.235839   1.044552
m->l      4:   0.228724   0.254321   1.111912
m->l      5:   0.248501   0.266627   1.072941
m->l      7:   0.252005   0.297201   1.179346
m->l      8:   0.275567   0.279873   1.015626
m->l      9:   0.272049   0.284632   1.046253
m->l     15:   0.276872   0.328499   1.186465
m->l     16:   0.278685   0.327017   1.173429
m->l     17:   0.314290   0.344236   1.095281
m->l     18:   0.320615   0.352390   1.099106
m->l    127:   0.628077   0.965992   1.538015
m->l    128:   0.634566   1.001855   1.578803
m->l    129:   0.997619   0.988396   0.990755
m->l    255:   1.117925   1.748253   1.563837
m->l    256:   1.136678   1.641906   1.444478
m->l    257:   5.223441   1.655042   0.316849
```

ZyX-I commented:
Updated regarding freeing list and list indexing. Though I mentioned that some of my benchmark code contains things it actually should not (specifically, allocated and freed on each iteration lists of two elements), so it needs to be fixed.

ZyX-I commented:
Updated benchmark results:

```
Test    Len    New        Old        Old/New
=============================================
alloc     0:   0.087877   0.072325   0.823025
alloc     1:   0.081558   0.070960   0.870056
alloc     2:   0.084648   0.077745   0.918451
alloc     3:   0.085255   0.080618   0.945610
alloc     4:   0.087998   0.104687   1.189652
alloc     5:   0.102352   0.094083   0.919210
alloc     7:   0.101435   0.120599   1.188929
alloc     8:   0.106456   0.114887   1.079197
alloc     9:   0.125766   0.130404   1.036878
alloc    15:   0.121160   0.170655   1.408509
alloc    16:   0.116585   0.160499   1.376669
alloc    17:   0.131462   0.158489   1.205588
alloc    18:   0.138567   0.163828   1.182302
alloc   127:   0.300016   0.738140   2.460335
alloc   128:   0.297527   0.750627   2.522887
alloc   129:   0.414556   0.747785   1.803821
alloc   255:   0.518358   1.347966   2.600454
alloc   256:   0.486582   1.360981   2.797023
alloc   257:   0.749532   1.373873   1.832974
1->$      0:   0.084023   0.083120   0.989253
1->$      1:   0.086216   0.083481   0.968277
1->$      2:   0.082848   0.081410   0.982643
1->$      3:   0.084478   0.086092   1.019106
1->$      4:   0.102198   0.081725   0.799673
1->$      5:   0.086454   0.094539   1.093518
1->$      7:   0.085787   0.092389   1.076958
1->$      8:   0.114656   0.081804   0.713473
1->$      9:   0.084673   0.083807   0.989772
1->$     15:   0.088395   0.082914   0.937994
1->$     16:   0.137070   0.083731   0.610863
1->$     17:   0.085790   0.087480   1.019699
1->$     18:   0.086235   0.080902   0.938157
1->$    127:   0.091158   0.093716   1.028061
1->$    128:   0.268846   0.098417   0.366072
1->$    129:   0.092408   0.087232   0.943988
1->$    255:   0.099424   0.090374   0.908976
1->$    256:   0.511325   0.091226   0.178411
1->$    257:   0.094770   0.090740   0.957476
1->^      0:   0.090562   0.088917   0.981836
1->^      1:   0.092504   0.091683   0.991125
1->^      2:   0.093668   0.098427   1.050807
1->^      3:   0.090920   0.098016   1.078047
1->^      4:   0.108733   0.091586   0.842302
1->^      5:   0.096598   0.092507   0.957649
1->^      7:   0.094974   0.095471   1.005233
1->^      8:   0.130676   0.093701   0.717048
1->^      9:   0.097546   0.093895   0.962572
1->^     15:   0.099166   0.091172   0.919388
1->^     16:   0.143332   0.097355   0.679227
1->^     17:   0.105921   0.094249   0.889805
1->^     18:   0.104156   0.095030   0.912381
1->^    127:   0.133787   0.101514   0.758773
1->^    128:   0.286983   0.108353   0.377559
1->^    129:   0.139526   0.102800   0.736780
1->^    255:   0.175886   0.098401   0.559459
1->^    256:   0.548609   0.098460   0.179472
1->^    257:   0.174739   0.101518   0.580969
1->/2     2:   0.097982   0.099614   1.016656
1->/2     3:   0.105240   0.101669   0.966068
1->/2     4:   0.113972   0.101282   0.888657
1->/2     5:   0.105076   0.100934   0.960581
1->/2     7:   0.106654   0.103226   0.967859
1->/2     8:   0.142102   0.101506   0.714318
1->/2     9:   0.101847   0.106942   1.050026
1->/2    15:   0.105189   0.118314   1.124775
1->/2    16:   0.151321   0.118118   0.780579
1->/2    17:   0.109668   0.116971   1.066592
1->/2    18:   0.109274   0.122119   1.117549
1->/2   127:   0.129072   0.177095   1.372064
1->/2   128:   0.284331   0.163905   0.576458
1->/2   129:   0.133691   0.167816   1.255253
1->/2   255:   0.142742   0.209094   1.464839
1->/2   256:   0.542819   0.201732   0.371638
1->/2   257:   0.148988   0.216335   1.452030
1<-$      1:   0.095618   0.095072   0.994290
1<-$      2:   0.102737   0.102857   1.001168
1<-$      3:   0.095212   0.093712   0.984246
1<-$      4:   0.094584   0.096989   1.025427
1<-$      5:   0.097975   0.094897   0.968584
1<-$      7:   0.096726   0.094988   0.982032
1<-$      8:   0.094383   0.093374   0.989310
1<-$      9:   0.097800   0.098734   1.009550
1<-$     15:   0.095335   0.108375   1.136781
1<-$     16:   0.093848   0.098442   1.048951
1<-$     17:   0.099392   0.099738   1.003481
1<-$     18:   0.097139   0.096766   0.996160
1<-$    127:   0.104103   0.124488   1.195816
1<-$    128:   0.105919   0.131427   1.240826
1<-$    129:   0.254964   0.127835   0.501385
1<-$    255:   0.107296   0.119212   1.111057
1<-$    256:   0.103915   0.117896   1.134543
1<-$    257:   0.279280   0.124203   0.444726
1<-^      1:   0.089175   0.088690   0.994561
1<-^      2:   0.090440   0.093852   1.037727
1<-^      3:   0.088677   0.096193   1.084757
1<-^      4:   0.092481   0.089766   0.970643
1<-^      5:   0.093882   0.098375   1.047858
1<-^      7:   0.094632   0.095191   1.005907
1<-^      8:   0.094768   0.098158   1.035772
1<-^      9:   0.097188   0.090432   0.930485
1<-^     15:   0.098337   0.099340   1.010200
1<-^     16:   0.096566   0.095619   0.990193
1<-^     17:   0.116227   0.094702   0.814802
1<-^     18:   0.106631   0.092030   0.863070
1<-^    127:   0.131802   0.115590   0.876997
1<-^    128:   0.129057   0.116625   0.903670
1<-^    129:   0.268974   0.120678   0.448660
1<-^    255:   0.180076   0.114468   0.635665
1<-^    256:   0.174982   0.114176   0.652501
1<-^    257:   0.396576   0.125487   0.316426
17<-$    17:   0.152554   0.128413   0.841754
17<-$    18:   0.151656   0.133889   0.882847
17<-$   127:   0.182495   0.147164   0.806400
17<-$   128:   0.180519   0.146344   0.810685
17<-$   129:   0.314597   0.145661   0.463008
17<-$   255:   0.183313   0.156032   0.851178
17<-$   256:   0.182045   0.154981   0.851333
17<-$   257:   0.376491   0.143580   0.381364
17<-^    17:   0.145671   0.119035   0.817150
17<-^    18:   0.144902   0.133080   0.918414
17<-^   127:   0.200682   0.137647   0.685896
17<-^   128:   0.215387   0.137438   0.638098
17<-^   129:   0.300873   0.136667   0.454235
17<-^   255:   0.239120   0.135324   0.565925
17<-^   256:   0.242867   0.139275   0.573462
17<-^   257:   0.433549   0.138226   0.318824
l->l      0:   0.099984   0.097434   0.974496
l->l      1:   0.109352   0.106215   0.971313
l->l      2:   0.102456   0.114838   1.120852
l->l      3:   0.125084   0.118448   0.946948
l->l      4:   0.125841   0.126522   1.005412
l->l      5:   0.149461   0.134634   0.900797
l->l      7:   0.143274   0.140859   0.983144
l->l      8:   0.144591   0.144817   1.001563
l->l      9:   0.161042   0.153360   0.952298
l->l     15:   0.163345   0.185336   1.134629
l->l     16:   0.170343   0.203898   1.196985
l->l     17:   0.200207   0.210618   1.052001
l->l     18:   0.202473   0.227028   1.121275
l->l    127:   0.434481   0.805504   1.853945
l->l    128:   0.412354   0.814209   1.974539
l->l    129:   0.618809   0.823154   1.330223
l->l    255:   0.782623   1.681700   2.148800
l->l    256:   0.779136   1.491395   1.914165
l->l    257:   1.511547   1.572375   1.040242
m->l      0:   0.136998   0.131534   0.960116
m->l      1:   0.141459   0.144451   1.021151
m->l      2:   0.141529   0.157846   1.115291
m->l      3:   0.150325   0.160834   1.069909
m->l      4:   0.156027   0.176737   1.132733
m->l      5:   0.172277   0.166497   0.966449
m->l      7:   0.177448   0.189685   1.068961
m->l      8:   0.176056   0.184486   1.047882
m->l      9:   0.205284   0.198222   0.965599
m->l     15:   0.201982   0.236048   1.168659
m->l     16:   0.198149   0.236473   1.193410
m->l     17:   0.237438   0.239933   1.010508
m->l     18:   0.242990   0.247810   1.019836
m->l    127:   0.451211   0.874145   1.937331
m->l    128:   0.492628   0.886857   1.800257
m->l    129:   0.688221   0.885110   1.286084
m->l    255:   0.819921   1.770554   2.159420
m->l    256:   0.827424   1.599653   1.933293
m->l    257:   0.997607   1.654314   1.658282
  -l-     0:   0.022980   0.020313   0.883943
  -l-     1:   0.022534   0.028148   1.249135
  -l-     2:   0.023742   0.039286   1.654705
  -l-     3:   0.024594   0.041437   1.684842
  -l-     4:   0.026456   0.049875   1.885206
  -l-     5:   0.032195   0.056503   1.755024
  -l-     7:   0.034117   0.068204   1.999121
  -l-     8:   0.034413   0.080128   2.328422
  -l-     9:   0.038168   0.085752   2.246699
  -l-    15:   0.046565   0.117183   2.516547
  -l-    16:   0.047724   0.130408   2.732545
  -l-    17:   0.052458   0.138526   2.640703
  -l-    18:   0.055536   0.132791   2.391080
  -l-   127:   0.157103   0.818254   5.208392
  -l-   128:   0.152894   0.831801   5.440377
  -l-   129:   0.201305   0.828444   4.115367
  -l-   255:   0.287399   1.588858   5.528405
  -l-   256:   0.289451   1.620139   5.597282
  -l-   257:   0.318472   1.620627   5.088758
l[0]      1:   0.044364   0.043979   0.991322
l[0]      2:   0.041034   0.044169   1.076400
l[0]      3:   0.042320   0.044234   1.045227
l[0]      4:   0.043018   0.043388   1.008601
l[0]      5:   0.048944   0.046792   0.956031
l[0]      7:   0.041388   0.043909   1.060911
l[0]      8:   0.044154   0.043696   0.989627
l[0]      9:   0.041584   0.043644   1.049538
l[0]     15:   0.042368   0.044868   1.059007
l[0]     16:   0.041677   0.044361   1.064400
l[0]     17:   0.045144   0.045256   1.002481
l[0]     18:   0.045662   0.045057   0.986750
l[0]    127:   0.057159   0.051400   0.899246
l[0]    128:   0.049031   0.052225   1.065142
l[0]    129:   0.059922   0.052325   0.873219
l[0]    255:   0.049186   0.051877   1.054711
l[0]    256:   0.050410   0.051539   1.022396
l[0]    257:   0.051912   0.053238   1.025543
l[$]      1:   0.046725   0.046782   1.001220
l[$]      2:   0.045594   0.045726   1.002895
l[$]      3:   0.044932   0.046918   1.044200
l[$]      4:   0.046472   0.048545   1.044608
l[$]      5:   0.046127   0.047875   1.037895
l[$]      7:   0.053326   0.045701   0.857012
l[$]      8:   0.053110   0.047692   0.897985
l[$]      9:   0.045116   0.045611   1.010972
l[$]     15:   0.045590   0.046802   1.026585
l[$]     16:   0.044659   0.047904   1.072662
l[$]     17:   0.049403   0.048618   0.984110
l[$]     18:   0.052336   0.051681   0.987485
l[$]    127:   0.052898   0.054848   1.036863
l[$]    128:   0.052774   0.054820   1.038769
l[$]    129:   0.054340   0.060098   1.105962
l[$]    255:   0.053033   0.056764   1.070352
l[$]    256:   0.054931   0.056349   1.025814
l[$]    257:   0.054104   0.055822   1.031754
l[/2]     1:   0.042525   0.048085   1.130747
l[/2]     2:   0.042654   0.045622   1.069583
l[/2]     3:   0.041363   0.044870   1.084786
l[/2]     4:   0.041325   0.043855   1.061222
l[/2]     5:   0.042173   0.045825   1.086596
l[/2]     7:   0.040547   0.044243   1.091153
l[/2]     8:   0.044835   0.053562   1.194647
l[/2]     9:   0.040934   0.049415   1.207187
l[/2]    15:   0.052001   0.058494   1.124863
l[/2]    16:   0.041670   0.058204   1.396784
l[/2]    17:   0.045393   0.059699   1.315159
l[/2]    18:   0.043566   0.061494   1.411514
l[/2]   127:   0.050653   0.102939   2.032239
l[/2]   128:   0.052449   0.100214   1.910694
l[/2]   129:   0.050519   0.103178   2.042360
l[/2]   255:   0.048980   0.148698   3.035892
l[/2]   256:   0.048759   0.134586   2.760229
l[/2]   257:   0.050252   0.148527   2.955644
```

Possible changes:

- Pop tests still faster with new code unless there is deallocation or list is small.
- Remove one element from beginning tests are still slower except for small tests which are almost the same.
- **Popping from end/removing from start of 17 elements tests are now always slower with new code.**
- Extending list with preallocated copy of itself test is still faster with new code unless doing so triggers double reallocation.
- For some reason all indexing tests are still mostly faster with new code, including indexing of start and end. Difference is less significant for start and end though.

ZyX-I commented:
Surprisingly, map() and filter() are always a bit slower for new implementation (much worse for filter()). Guess that lots of levels of stack needed to evaluate second argument successfully erase away caches, after what new implementation appears to be worse because it needs to check list size.

ZyX-I commented:
Sorting is always a bit faster with new implementation.

ZyX-I commented:
While optimising list found a bug: there may be crash if list item edited by map() gets removed by map() expression. E.g. the following always crashes:

    vim -i NONE -u NONE -N --cmd 'let l = ["abc"]|call map(l, "remove(l, 0)[0]")' --cmd qa

After my changes to list implementation I for some reason cannot trigger a crash like this, but problem should be there still.

ZyX-I commented:
vim/vim#2652

justinmk commented:
Test    Len    New        Old        Old/New
    alloc     0:   0.087877   0.072325   0.823025

What is old/new column supposed to represent? `0.087877/0.072325 = 1.21`

ZyX-I committed:
eval: Refactor map/filter implementation

Intended as optimization, but did not achieve anything significant.

ZyX-I committed:
functests: Make output of print_snapshot wrap highlight attributes

ZyX-I committed:
functests: Dump unknown colors as hex strings

ZyX-I commented:
New map implementation is faster if I switch how lists are populated: normally it is `map(ll, 'copy(l)')` and it most likely makes allocator put list values nearby (especially since benchmark build uses jemalloc). If I instead use

```VimL
                    call map(ll, "[]")
                    call map(l, "map(ll, 'add(v:val, '.string(v:val).')')")
```

I get almost half a minute for list with 257 elements, but ~5.0 seconds on master and ~4.4 seconds with new implementation.

ZyX-I commented:
(I mean, half a minute for running benchmark itself, five and four seconds are benchmark results.)

ZyX-I commented:
Gcc complains about “missing initializer” for `{ .attr = … }` style initializer. I thought it is OK. And my gcc does not complain about that.

ZyX-I committed:
build: Use more specific feature macros

This was actually an attempt to fix local clang warnings from glibc-2.25
macros, specifically:

1. `-Wconversion` on fpclassify() macros.
2. `-Wgnu-statement-expression` from `assert()` macros.

Though after some attempts it appeared that

1 is not actually fixable without making clang ignore warnings from
  math.h (should already be doing that, but does not for some reason) or
  making it pretend being gcc-4.4 or higher and not gcc-4.2 (in the
  former case glibc uses __builtin_fpclassify which is actually
  defined).
2 is also not fixable withoun using `-ansi` (or defining
  `__STRICT_ANSI__` or undefining `__GNUC__`). Though this should be
  fixed by a newer libc: it [uses `__extension__`][1] to mask warnings.

[1]: https://sourceware.org/git/?p=glibc.git;a=blob;f=assert/assert.h;h=5468c16c2c4e354e86454386c39a049daa130523;hb=23158b08a0908f381459f273a984c6fd328363cb#l105

dgel commented:
Hi! I'm sorry for just dropping into this pull request, since I'm not a neovim dev and don't really follow the existing code. I just had a few questions about this, since similar comparisons are often done in c++ for std::list vs std::vector. Do I understand correctly that the old implementation was a doubly-linked list of basic types with some sort of caching mechanism? and the new one is a dynamic array implementation?

- You mention that half the lists are empty, but also that you always allocate 4 elements. Do you mean that you allocate 4 elements upon inserting the first element? 
- If most vectors are small, does it make sense to use some form of small_vec implementation to avoid allocating for very small vectors?
- inserting at the end being faster for lists seems odd, but maybe it's because you're comparing small lengths only
- in c++, std::vector doesn't shrink until requested. Perhaps this doesn't make sense for vimscript, but it's worth considering deallocating less often than you allocate (double when capacity is full, shrink vector when less than half or one third capacity?)
- why can extending a list with itself reallocate twice?

ZyX-I commented:
0. Old implementation is a double-linked list. Accessing a certain index (as index) leaves that in a cache. New variant is dynamically `realloc()`ated array with size equal to one of the powers of two.
1. I am allocating 4 elements upon list allocation. This does not increase memory usage too much and makes single `malloc()` call good for 90% of lists.
2. I do not know C++ and can’t say what is “some form of small_vec implementation”. Google knows about rust’s, but I do not see how I can use stack, list is there for VimL implementation which needs keeping pointers around for undetermined time, for internal uses there are normally other equivalents.
3. This were my thoughts as well. Inserting at the end ought to be faster as long as there is free space: you don’t need to allocate and you need to write to less places: with double linked list places to write are

   1. List length saved in structure.
   2. Pointer to the last element.
   3. Pointer in the element itself. Two unless using calloc().
   4. Value in the element itself.

   With vectors that would be

   1. List length.
   2. Value of the element.

   And that’s all (unless reallocation happens). How it ended up slower dunno.
4. Removes are relatively rare and it is not like there is a possibility of keeping “vector size is a power of two” rule without shrinking when more then 1/2 of capacity filled. 1/4 may be used, but I do not find optimizing that a priority.
5. Do not remember why I thought it was happening. Should actually be another explanation: e.g. with 255/256 elements it needs to copy 256 elements when reallocation happens, with 257 elements it needs to copy 512 elements when reallocation happens because 512 elements were already allocated, but 257*2 = 514 do not fit there.

dgel commented:
Hi, thanks for replying =) Possibly I'm misunderstanding and you're doing it already, but would it make sense to not allocate at all until an element is added to the vector? 

By a small_vector I meant a vector that has an inline buffer (union with the vector's pointer to the buffer, size and capacity) where on small sizes it uses the inline buffer instead of allocating. This may be total overkill / counter productive however.

ZyX-I commented:
@dgel How do you imagine not allocating? I need a editable object so that two variables holding the same empty vector will get “changed” by putting a value to it.

About small: maybe some code? I kind of have something like this now: allocating a list will allocate space for structure which has four list items among other things. There are no unions though, vector with preallocated items is supposed to be compatible with vector with no preallocated items (i.e. any function which operates on the pointer to vector without preallocated items should also work with pointer to vector with preallocated items) as long as changes made to the vector do not trigger some kind of reallocation. So in place of code checking size against number of preallocated items and selecting where to get values from pointer to the items array is used, which incidentally points to the memory just after the memory where pointer itself is stored. Less space efficient, but easier to work with: I actually use this feature to have vectors with different number of preallocated items (Vim has two lists with different number of preallocated items: for group captures when pattern matching and for user function arguments), and also still be able to easily switch back to vector without any preallocated items at all.

dgel commented:
Ahh yeah it looks like I was misunderstanding. I thought a `struct { Data* buffer; size_t cap; size_t len}` was being passed around directly, but it seems you're passing a pointer to something like `struct {Data* buffer, size_t cap; size_t len; Data initial[4]; }`? Or perhaps you always keep the data inline and grow the entire struct instead of holding a pointer to a separate buffer. Either way there's no real advantage to making it a union other than to try to save 16 bytes per vector.

Thanks for talking about this! It's nice to learn about how such lists are implemented in different places. Perhaps I can find an English-language equivalent of that article about different list implementations

ZyX-I commented:
I can’t grow the entire struct ever. First, that makes vector not compatible with vector without initial elements. Second, I do need a never-changing pointer to hold and right now nearly the same struct you wrote is a part of another struct which additionally has data for GC and some other technical thingies. I considered the variant of having `Data buffer[]` at the end and grow it, but it neither allows using existing kvec implementation nor has an advantage of less amount of by-pointer memory accesses. By the way, I have seen exactly zero languages where `cap` and `size` are stored in the same memory block as `buffer` side from some optimizations in Ruby (does use `union`, but in Ruby `Data` is actually `Data *`, in Neovim that is complex `struct` with actual data most of time; also once union is overfilled data array is no longer in the same memory block because of it being accessed via a pointer).

(About article: I have the data collected, but did not quite like how I structured it yet, so did not publish it so far. So you would not have that in Russian either yet.)

ZyX-I committed:
Merge branch 'master' into list-reimplement

ZyX-I committed:
oldtests: Make sure that targets order is correct

With this `make -jN` is somewhat working (if you are lucky, tests still do share 
common filenames). Better though is that there are no implicit dependencies now.

ZyX-I committed:
oldtests: Allow specifying timeout (in minutes) via env var

ZyX-I committed:
oldtests: Allow running old tests in parallel

This actually makes sense only on developer machines and adds overhead (copying 
all test files for each test run), but not much of that.

dundargoc commented:
Needs rebase.
